### DISCUSSION TITLE
kitten icat: Failed to open controlling terminal

### DISCUSSION THREAD

**Discussion by iambumblehead** [kovidgoyal/kitty#7275] [Q&A] [answered]
Is there a way to use kitten icat inside an ncurses application?

Using `kitten icat` inside an ncurses application here gives the error `Failed to open controlling terminal`
```bash
kitten icat --align left "$img_path"
```

Per [this thread](https://github.com/kovidgoyal/kitty/issues/2426#issuecomment-711095305) I've also tried the following commands which result in different errors,
```bash
kitten @ --to=$KITTY_LISTEN_ON icat --align left "$img_path"
# Error: icat is not a known subcommand for @. Use --help to get a list of valid subcommands.
kitten icat --align left "$img_path" @ --to=$KITTY_LISTEN_ON
# Error: Failed to open controlling terminal with error: open /dev/tty: device not configured
```

thanks in advance for any help or advice

**Comment by kovidgoyal** [comment]
On Wed, Mar 27, 2024 at 08:18:32PM -0700, iambumblehead wrote:
> Is there a way to use kitten icat inside an ncurses application?

Yes, just run it, and make sure your application is not doing any IO
with /dev/tty while running it.

> 
> Using `kitten icat` inside an ncurses application here gives the error `Failed to open controlling terminal`
> ```bash
> kitten icat --align left "$img_path";;
> ```

That's not an ncurses application, thats probably nvim, which works in
client server mode and has no access to the tty. If you want to display
images in nvim use images.nvim

> 
> Per [this thread](https://github.com/kovidgoyal/kitty/issues/2426#issuecomment-711095305) I've also tried the following commands which result in different errors,
> ```bash
> kitten @ --to=$KITTY_LISTEN_ON icat --align left "$img_path";;
> # Error: icat is not a known subcommand for @. Use --help to get a list of valid subcommands.
> kitten icat --align left "$img_path" @ --to=$KITTY_LISTEN_ON ;;
> # Error: Failed to open controlling terminal with error: open /dev/tty: device not configured
> ```

Remote control cant be used to display images,
see https://sw.kovidgoyal.net/kitty/remote-control/

**Reply by iambumblehead** [reply]
the parent application is vifm and basically this same issue is discussed at stackoverflow here https://stackoverflow.com/questions/76024179/issue-with-previewing-images-in-vifm-on-wayland-with-kitty-on-arch-linux

**Comment by kovidgoyal** [comment]
On Wed, Mar 27, 2024 at 08:29:00PM -0700, iambumblehead wrote:
> the parent application is vifm and basically this same issue is discussed at stackoverflow here https://stackoverflow.com/questions/76024179/issue-with-previewing-images-in-vifm-on-wayland-with-kitty-on-arch-linux

Then you need to get vifm to allow you to run processes with access to
the terminal device. Or get it to suppor thte kitty image protocol
natively.

**Comment by kovidgoyal** [comment]
On Wed, Mar 27, 2024 at 08:29:00PM -0700, iambumblehead wrote:
> the parent application is vifm and basically this same issue is discussed at stackoverflow here https://stackoverflow.com/questions/76024179/issue-with-previewing-images-in-vifm-on-wayland-with-kitty-on-arch-linux

Oh and that stack overflow answer is incomplete. You dont need a terminal
if you can pass the window dimensions manually, as demonstrated by

zsh -c 'setsid kitten icat --use-window-size 100,30,3000,2000 --place ***@***.*** --transfer-mode=file logo/kitty.png'

You will of course need to replace the various dimensions and the
filename. But as I said in my previous reply, its easier just to use the
tty.

**Reply by kovidgoyal** [reply]
And see https://github.com/kovidgoyal/kitty/commit/399a9d65d2be6d0021b91ccc465bcfc0c9eb45db

**Reply by iambumblehead** [reply]
Thanks for replying and helping me.  The approach suggested at the vifm issue linked from stackoverflow is working here. 'Will try setting window-size and removing /dev/tty redirections next,
```bash
kitten icat \
  --place "${img_wh}@${img_tl}" \
  --align left \
  --transfer-mode=stream "$img_path" >/dev/tty </dev/tty
```

<img width="400" src="https://github.com/kovidgoyal/kitty/assets/2058705/15dddac4-365e-4058-bc29-9219f9a5a6d9">

**Reply by iambumblehead** [reply]
after updating kitty; `--use-window-size` is working without errors

**Reply by iambumblehead** [reply]
removing the /dev/tty redirections (either one or both) results in an error
```bash
kitten icat \
  --use-window-size "${kicat_wh_cells/x/,},${kicat_wh_pixels/x/,}" \
  --place "${img_wh}@${img_tl}" \
  --align left \
  --transfer-mode=stream "$img_path"
```

error
```bash
Error: The --place option can only be used with a single image, not 2
```

**Comment by kovidgoyal** [comment]
On Wed, Mar 27, 2024 at 10:03:16PM -0700, iambumblehead wrote:
> removing the /dev/tty redirections (either one or both) results in an error
> ```bash
> kitten icat \
>   --use-window-size "${kicat_wh_cells/x/,},${kicat_wh_pixels/x/,}" \
>   --place "${img_wh}@${img_tl}" \
>   --align left \
>   --transfer-mode=stream "$img_path"
> ```
> 
> error
> ```bash
> Error: The --place option can only be used with a single image, not 2
> ```

Add --stdin=no

**Reply by iambumblehead** [reply]
> Add --stdin=no

error is not seen after adding that, however image is no longer rendered --instead long text fills the screen

<img width="400" alt="Screen Shot 2024-03-27 at 10 16 05 PM" src="https://github.com/kovidgoyal/kitty/assets/2058705/84c06423-11b4-4bd5-b791-4937c22aabb7">

**Reply by iambumblehead** [reply]
updating the command to use `--transfer-mode=file` but no change occurred

**Comment by kovidgoyal** [comment]
On Wed, Mar 27, 2024 at 10:16:59PM -0700, iambumblehead wrote:
> > Add --stdin=no
> 
> error is not seen after adding that, however image is no longer rendered --instead long text fills the screen

that will be because vifm is mangling the escape codes it gets from
icat before passing them to the tty. So stick to using the %N mode with /dev/tty.

**Reply by iambumblehead** [reply]
It is agreeble if the discussion is closed as "resolved" however one point of mention; the same script renders sixel image format from inside vifm plus <strike>other terminal emulators without /dev/tty redirections</strike> It turns out the sixel function below also displays noise when used with vifm+iterm2
```bash
image_to_sixel_magick () {
    img_path=$1
    img_wh=$3

    export MAGICK_OCL_DEVICE=true
    if [[ -n "$is_cmd_magick" ]]; then
        magick \
            -background "rgba(0,0,0,0)" \
            "$img_path" \
            -geometry "$img_wh" \
            sixel:-
        echo ""
    elif [[ -n "$is_cmd_convert" ]]; then
        convert \
            -channel rgba \
            -background "rgba(0,0,0,0)" \
            -geometry "$img_wh" \
            "$img_path" \
            sixel:-
        echo ""
    else
        fail "$msg_cmd_not_found_magickany"
    fi
}
```

**Reply by kovidgoyal** [reply, marked as answer]
Then you should ask the vifm developers to pass through kitty graphics protoco escape codes since they do that for sixel escape codes already.

**Reply by iambumblehead** [reply]
yes, kitty is working correctly here and the issue originates at vifm.

This issue occurs when vifm is configured to provide both top and left position to the image viewer script. I assumed this was a bug and reported it to vifm. 

The developer explained that, when vifm is configured to provide both top and left position (which the script uses to define `--use-window-size "${kicat_wh_cells/x/,},${kicat_wh_pixels/x/,}"`) it makes assumptions about the returned value and escapes it.

There is no issue at kitty and the related vifm discussion is here https://github.com/vifm/vifm/issues/731#issuecomment-2026259316

Thank you for facilitating and helping me :)
